src/transformers/models/gpt2/modeling_gpt2.py

```diff
@@ -192,6 +210,62 @@ def _attn(self, query, key, value, attention_mask=None, head_mask=None):
             attn_weights = attn_weights + attention_mask
 
         attn_weights = nn.Softmax(dim=-1)(attn_weights)
+
+        # Downcast (if necessary) back to V's dtype (if in mixed-precision) -- No-Op otherwise
+        attn_weights = attn_weights.type(value.dtype)
+        attn_weights = self.attn_dropout(attn_weights)
+
+        # Mask heads if we want to
+        if head_mask is not None:
+            attn_weights = attn_weights * head_mask
+
+        attn_output = torch.matmul(attn_weights, value)
+
+        return attn_output, attn_weights
+
+    def _upcast_and_reordered_attn(self, query, key, value, attention_mask=None, head_mask=None):
+        # Use `torch.baddbmm` (a bit more efficient w/ alpha param for scaling -- from Megatron-LM)
+        bsz, num_heads, q_seq_len, dk = query.size()
+        _, _, k_seq_len, _ = key.size()
+
+        # Preallocate attn_weights for `baddbmm`
+        attn_weights = torch.empty(bsz * num_heads, q_seq_len, k_seq_len, dtype=torch.float32, device=query.device)
+
+        # Compute Scale Factor
+        scale_factor = 1.0
+        if self.scale_attn_weights:
+            scale_factor /= float(value.size(-1)) ** 0.5
+
+        if self.scale_attn_by_inverse_layer_idx:
+            scale_factor /= float(self.layer_idx + 1)
```

Comment:
@siddk would you mind explaining why this is an option? The only place I've seen something similar before is in https://github.com/NVIDIA/Megatron-LM/blob/e156d2fea7fc5c98e645f7742eb86b643956d840/megatron/model/transformer.py#L157-L158 but it gets renormalized in `FusedScaleMaskSoftmax`. Effectively what I believe Megatron-LM does is
```
softmax ( QV / sqrt(d) / layer_id * layer_id)
```
What I believe this PR to be doing:
```
softmax( QV / sqrt(d) / layer_id)
```

Reply:
Yeah so this was inspired by the Megatron code - the dividing by layer index was a heuristic to prevent overflow in FP16, which we ran into a few times training the Mistral GPT-2 small and medium models.

The `baddbmm` in the lines below actually applies the scaling value to the keys prior to the QK multiply. We chose not to re-multiply by layer_idx -- it changes the semantics a bit, but doesn't affect performance.

Reply:
should this be changed for fp32 and bf16 cases where there will be no overflow?

Reply:
So we actually hit overflow without this in FP32 for one of our runs. I think leaving this as a flag users can turn on might be ok? The flag/docs are pretty clear about what it does. What do you think @stas00?

Reply:
my apologies - I missed that it was conditional on the user configurable flag - all is good!

Just following the discussion I find it's always good to leave a good commentary for why certain things are done to avoid questions in the future - no need to do anything at the moment - just a suggestion for the future.

Reply:
Absolutely - thanks for the advice Stas, will keep it in mind moving forward!

@thomasw21 - does this answer your question?

Reply:
Yes thank you! Thank you for confirming this had a different behaviour to Megatron-LM.